nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1215,6 +1231,79 @@ impl DataStore {
         }
     }
 
+    /// Set an instance's CPU count and memory size to the provided values,
+    /// within an existing transaction.
+    async fn instance_set_size_on_conn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<Error>,
+        authz_instance: &authz::Instance,
+        ncpus: InstanceCpuCount,
+        memory: ByteCount,
+    ) -> Result<(), diesel::result::Error> {
+        use db::schema::instance::dsl as instance_dsl;
+
+        // Do not allow setting sizes of running instances to ensure that if an
+        // instance is running, its resource usage matches what we record in the
+        // database.
+        const OK_TO_SET_SIZE_STATES: &'static [InstanceState] = &[
+            InstanceState::NoVmm,
+            InstanceState::Failed,
+            InstanceState::Creating,
+        ];
+
+        let maybe_old_sizes = instance_dsl::instance
+            .filter(instance_dsl::id.eq(authz_instance.id()))
+            .filter(instance_dsl::time_deleted.is_null())
+            .select((instance_dsl::ncpus, instance_dsl::memory))
+            .first_async::<(InstanceCpuCount, ByteCount)>(conn)
+            .await;
```

Comment:
hmm, we are now fetching fields from the instance record in a bunch of different places in the `instance_reconfigure` method...perhaps it would be nicer if we just did one `instance_fetch` query and then passed an `&Instance` into both this function and `instance_set_boot_disk` to perform the validation? Not sure how big a deal this is, though.

Reply:
Alternatively, it occurs to me that we could just fold this into the update query, by adding a
```rust
.filter(instance_dsl::ncpus.ne(ncpus).or(instance_dsl::memory.ne(memory))
``` 
clause to make the update conditional on there actually being a change?

We would then have to handle that case in `NotUpdatedButExists` as well as the "instance is in invalid state" case, but I think that's probably fine and might be a bit more efficient than doing a query, checking the results in Nexus, and then doing another query.

Reply:
ah, i'd thought about that and written it off because i'd also been thinking about the caller checking instance state as well, but `Instance` has `.state` as well, so the state check can stay in these helpers while `&Instance`-to-be-updated comes from the caller. and i wasn't enthused about loading the instance again in `sic_set_boot_disk`, but we already have that in the saga parameters!

i'll give that a try, especially because it means we don't even have to do these prepatory queries over and over. adding `.filter` for the fields not matching would be nice but, yeah, that's extra queries.

even though update is probably going to be relatively infrequent, i'd like to make this not awful to read :) i'll see how it goes.

Reply:
> `.filter()` trip report: you might think the boot disk filter stanza would look something like:
> 
> ```rust
> .filter(instance_dsl::boot_disk_id.ne(boot_disk_id))
> ```
> 
> but `NULL` compares `false` with everything, so this skips updating the boot disk if either of the prior or new `boot_disk_id` are None. okay. so the real precise filter would be more like:
> 
> ```rust
> .filter(
>     instance_dsl::boot_disk_id.ne(boot_disk_id)
>         .or(instance_dsl::boot_disk_id.is_null().and(boot_disk_id.is_some()))
>         .or(instance_dsl::boot_disk_id.is_not_null().and(boot_disk_id.is_none()))
> )
> ```
> 
> but this starts not working for "the types are not what diesel wants to see" reasons.

so...you're gonna hate me for this, but...couldn't we just do, essentially,
```rust
if boot_disk_id.is_some()
    // query...
    .filter(
        instance_dsl::boot_disk_id.ne(boot_disk_id)
            .or(instance_dsll::boot_disk_id.is_null())
    )
    // ... more query ...
} else {
    // query...
    .filter(
        instance_dsl::boot_disk_id.ne(boot_disk_id)
            .or(instance_dsll::boot_disk_id.is_not_null())
    )
    // ... more query ...
}
```

Reply:
actually if the types work out on
```rust
let boot_disk_filter = if boot_disk_id.is_some() {
    instance_dsl::boot_disk_id.ne(boot_disk_id).or(instance_dsl::boot_disk_id.is_null())
} else {
    instance_dsl::boot_disk_id.ne(boot_disk_id).or(instance_dsl::boot_disk_id.is_not_null())
}

let r = diesel::update(instance_dsl::instance)
    .filter(boot_disk_filter)
    // .. and the rest ..
```
that's totally fine.

Reply:
i suspect those are different types, but you can probably get around it by using `.boxed()` on the query in the appropriate place

Reply:
yeah this looks pretty reasonable now, thanks! https://github.com/oxidecomputer/omicron/pull/6774/files#diff-a234811eb5ee38fdcf626cd6514deb7dbfc8b9e2067cfa3b4a6f111d72a07e53R1171